---
title: "Recount3 Analysis Time"
output: html_notebook
---

# Intro
After the events of recount3/README.md we can now start the analysis
```{r}
library(tidyverse)
library(matrixStats)
library(ggrepel)
base_dir <- '~/data/metaRPE_RSE/'
load('../data/rse_gene_select.Rdata')

# set XY ratio
colData(rse_gene_select)$XYratio <- (colData(rse_gene_select)$"recount_qc.aligned_reads%.chrx" + 0.01) / (colData(rse_gene_select)$"recount_qc.aligned_reads%.chry" + 0.01)

# make merged cell line / owner field
colData(rse_gene_select)$CL_Owner <- paste(colData(rse_gene_select)$cell_line, colData(rse_gene_select)$Owner, sep = '_')

# load in gff
gff <- rtracklayer::import.gff(paste0(base_dir, '/human/annotations/gene_sums/human.gene_sums.G029.gtf.gz')) %>% as_tibble()
```

# Metadata
```{r}
colData(rse_gene_select) %>% as_tibble() %>% select(Sample:DrugSimple) %>% janitor::remove_empty() %>% DT::datatable()
```
# VST Transform and Build DESeq2 Object
```{r, include=FALSE}
dds <- DESeqDataSetFromMatrix(countData = assay(rse_gene_select, "counts"), colData = colData(rse_gene_select), design = ~Class + CL_Owner)
DESeq2Table <- DESeq(dds, parallel = T)
vst <- varianceStabilizingTransformation(DESeq2Table)
colnames(vst) <- colData(dds)$Sample
```

# PCA
## Plot by Owner (color) and Class (shape)
This looks quite good. Essentially PC1 is separating controls and degeneration (with two sort of exceptions) Samples. PC2 separates Karla (trysin) and everything else. 
```{r, fig.width=5, fig.height=4}
#pca 
library(matrixStats)
ntop = 2000
Pvars <- rowVars(assay(vst))
select <- order(Pvars, decreasing = TRUE)[seq_len(min(ntop, 
                                                      length(Pvars)))]
PCA <- prcomp(t(assay(vst)[select, ]), scale = F)
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
dataGG <- cbind(PCA$x, colData(vst)) %>% data.frame()
ggplot(dataGG, aes(PC1, PC2, color=Owner, shape=Class)) +
  ggrepel::geom_label_repel(aes(label = Sample)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  #facet_wrap(~line) +
  cowplot::theme_cowplot()
ggplot(dataGG, aes(PC3, PC4, color=Owner, shape=Class)) +
  ggrepel::geom_label_repel(aes(label = Sample)) +
  geom_point(size=3) +
  xlab(paste0("PC3: ",percentVar[3],"% variance")) +
  ylab(paste0("PC4: ",percentVar[4],"% variance")) + 
  ggrepel::geom_label_repel(aes(label = Sample)) +
  cowplot::theme_cowplot()


```

## Plot by Cell Line (color) and Class (shape)

```{r, fig.width=5, fig.height=4}
ggplot(dataGG, aes(PC1, PC2, color=cell_line, shape=Class)) +
  ggrepel::geom_label_repel(aes(label = Sample)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  #facet_wrap(~line) +
  cowplot::theme_cowplot() +
  scale_color_manual(values = pals::glasbey() %>% unname())

ggplot(dataGG, aes(PC3, PC4, color=cell_line, shape=Class)) +
  ggrepel::geom_label_repel(aes(label = Sample)) +
  geom_point(size=3) +
  xlab(paste0("PC3: ",percentVar[3],"% variance")) +
  ylab(paste0("PC4: ",percentVar[4],"% variance")) + 
  cowplot::theme_cowplot() +
  scale_color_manual(values = pals::glasbey() %>% unname())
```


```{r, fig.height=4, fig.width=5}
library(ComplexHeatmap)
SampleDists <- dist(t(assay(vst)[select, ]))
SampleDistMatrix <- as.matrix(SampleDists) %>% scale(., center= FALSE)
colnames(SampleDistMatrix) <- colData(vst)$Sample
row.names(SampleDistMatrix) <- colData(vst)$Sample
ha_column = HeatmapAnnotation(df = data.frame(Owner = colData(vst)$Owner,
                                              Class = colData(vst)$Class),
                              col = list(Owner = c("karla" = pals::alphabet2(20)[3] %>% unname(),
                                                   "davide" = pals::alphabet2(20)[6] %>% unname(), 
                                                   "qin" = pals::alphabet2(20)[9] %>% unname(),
                                                   "ruchi" = pals::alphabet2(20)[12] %>% unname(),
                                                   "dominik" = pals::alphabet2(20)[16] %>% unname()),
                                         Class = c("Control" = viridis::magma(10)[3], 
                                                   "Degeneration" = viridis::magma(10)[7])))
max_dist <- 1.7
Heatmap(SampleDistMatrix, 
        top_annotation = ha_column,name = 'Distance',
        col = circlize::colorRamp2(seq(0,max_dist,0.1), 
                                   colors = viridis::viridis(n = length(seq(0,max_dist,0.1)))))
```

```{r}
res <- results(DESeq2Table, contrast = c('Class','Control','Degeneration'))

plotMA(res, ylim=c(-5,5))
```
# Top 2000 Diff Gene

```{r}
res %>% 
  as_tibble(rownames = 'gene_id') %>% 
  left_join(gff) %>% 
  filter(padj < 1e-5, abs(log2FoldChange) > 1) %>% 
  head(2000) %>% 
  arrange(padj) %>% 
  DT::datatable()
```

# Volcano
Values on the right are more highly expressed in control iRPE vs degeneration iRPE
```{r, fig.width=5}
volcano_maker <- function(df, title="Volcano Plot", pvalue='pvalue', padj='padj', logFC='log2FoldChange', gene_list = ''){
  df <- df[!is.na(df$pvalue),]
  print(dim(df))
  df$Class <- 'Not Significant'
  df$Class[df[,'padj'] < 0.01 & df[,'log2FoldChange'] %>% abs() > 2] <- "FDR < 0.01 &\nabs(log2FC) > 2"
  df$GeneT <- df$Gene
  df$Gene[!df$Gene %in% gene_list] <- ''
  
  plot <- ggplot(data=df,aes(label=Gene, x = log2FoldChange, y = -log10(pvalue))) + 
    geom_point(aes(colour=Class)) +
    scale_colour_manual(values=c("darkred","grey")) + 
    cowplot::theme_cowplot() +
    geom_label_repel(data=bind_rows(subset(df, log2FoldChange > 2 & baseMean > 500) %>% arrange((padj)) %>% head(10), 
                                    subset(df, log2FoldChange < -2 & baseMean > 500) %>% arrange((padj)) %>% head(10)), 
                     aes(label=GeneT)) +
    xlab('logFC') + ylab('-log10(p value)') +
    ggtitle(title) + cowplot::theme_cowplot() +
    geom_vline(aes(xintercept=-2),linetype="dotted") +
    geom_vline(aes(xintercept=2),linetype="dotted") 
  plot
}
volcano_maker(df =res %>% as_tibble(rownames = 'gene_id') %>% left_join(gff) %>% mutate(Gene = gene_name)) + coord_cartesian(xlim = c(-5,5), title= 'iRPE Ctrl vs Degeneration')
```



# Dot Plots of Select Genes
```{r, fig.width=5}
gene_counts <- assay(dds, slot = 'counts')
colnames(gene_counts) <- colData(dds)$Sample
plot_genes <- c(res %>% as_tibble(rownames = 'gene_id') %>% left_join(gff) %>% filter((log2FoldChange) > 2, baseMean > 500) %>% arrange((padj)) %>% head(10) %>% pull(gene_name), 
                res %>% as_tibble(rownames = 'gene_id') %>% left_join(gff) %>% filter((log2FoldChange) < -2, baseMean > 500) %>% arrange((padj)) %>% head (10) %>% pull(gene_name))
gene_counts %>% as_tibble(rownames = 'gene_id') %>% 
  pivot_longer(-gene_id) %>% 
  left_join(gff) %>% 
  left_join(colData(dds) %>% as_tibble(), by = c('name' = 'Sample')) %>% 
  filter(gene_name %in% plot_genes) %>% 
  ggplot(aes(x=Class, y = log2(value), color = Owner)) + 
  geom_boxplot(color = 'black', outlier.colour = NA) +
  ggbeeswarm::geom_quasirandom() + 
  facet_wrap(~gene_name) +
  cowplot::theme_cowplot() +
  coord_flip()
```

# Heatmap of Bharti RPE Markers
```{r, fig.width=5, fig.height=14}
rpe_markers <- scan('~/git/eyeMarkers/lists/bharti_rpe_markers.txt', what = 'character')
source('../src/heatmap.R')
make_hm(dds, rpe_markers, min_val = -5, max_val = 5)
```

# GO Enrichment

Control vs Degeneration
```{r, fig.width=5}
library(clusterProfiler)
library(enrichplot)
diff_genes <- res %>% as_tibble(rownames = 'gene_id') %>% left_join(gff) %>% filter(padj < 0.01, abs(log2FoldChange) > 2) 
eg_diff_genes <- bitr(diff_genes$gene_name, fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
eg_diff_genes <- diff_genes %>% left_join(., eg_diff_genes, by = c('gene_name' = 'SYMBOL'))
eg_universe = bitr(res %>% as_tibble(rownames = 'gene_id') %>% left_join(gff) %>% pull(gene_name), fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
ego <- enrichGO(gene          = eg_diff_genes$ENTREZID,
                universe      = eg_universe$ENTREZID,
                OrgDb         = org.Hs.eg.db,
                ont           = "all",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
                readable      = TRUE)
p1 <- dotplot(ego, showCategory=30) + ggtitle("Dotplot for GO")
p1
```

## "CNET" Plot
Relationships between related GO terms with shared genes
```{r, fig.width=9, fig.height=6}
geneList <- eg_diff_genes$log2FoldChange
names(geneList) <- eg_diff_genes$gene_id
cnet <- cnetplot(ego, foldChange = geneList, showCategory =20 ) + scale_color_viridis_c(name = 'log2(FoldChange)')
cnet
```
## Enriched GO Term Relationships
Shows the "sets" of GO terms to increase understanding of what the diff gene sets are doing. There are many related GO terms.
```{r, fig.width=6}
ego_pair <- pairwise_termsim(ego) 
go_term_relationships <- emapplot(ego_pair,layout="kk") #+ ggtitle("Enriched GO Term Relationships")
go_term_relationships
```
## Table
```{r}
egoOF_DB@result %>% as_tibble() %>%  filter(p.adjust < 0.05) %>%  DT::datatable()
```

# Wikipathways
```{r, fig.width=4, fig.height=7}
# system("wget https://wikipathways-data.wmcloud.org/current/gmt/wikipathways-20211210-gmt-Homo_sapiens.gmt")
wp2gene <- read.gmt('../wikipathways-20211210-gmt-Homo_sapiens.gmt')
wp2gene <- wp2gene %>% tidyr::separate(term, c("name","version","wpid","org"), "%")
wpid2gene <- wp2gene %>% dplyr::select(wpid, gene) #TERM2GENE
wpid2name <- wp2gene %>% dplyr::select(wpid, name) #TERM2NAME
ewp <- enricher(eg_diff_genes$ENTREZID,
                TERM2GENE = wpid2gene,
                TERM2NAME = wpid2name,
                pvalueCutoff = 0.1)
ewp_plot <- dotplot(ewp, showCategory=30) + ggtitle("Dotplot for WikiPathways")
ewp_plot
```

## Table
```{r}
ewp <- setReadable(ewp, OrgDb =  org.Hs.eg.db, keyType = 'ENTREZID')
ewp@result %>% DT::datatable()
```


# GSEA
```{r, fig.width=5, fig.height=5}
all_genes <- bitr(res %>% as_tibble(rownames = 'gene_id') %>% left_join(gff) %>% pull(gene_name), fromType="SYMBOL", toType="ENTREZID", OrgDb="org.Hs.eg.db")
all_genes <- all_genes %>% left_join(res %>% as_tibble(rownames = 'gene_id') %>% left_join(gff), by = c('SYMBOL' = 'gene_name'))
logFC <- all_genes$log2FoldChange
names(logFC) <- all_genes$ENTREZID
logFC <- na.omit(logFC)
logFC = sort(logFC, decreasing = TRUE)
gse <- gseGO(geneList=logFC,
             ont ="ALL",
             keyType = "ENTREZID",
             pvalueCutoff = 0.05,
             OrgDb = org.Hs.eg.db,
             pAdjustMethod = "BH")
dotplot(gse, showCategory=10, split=".sign") + facet_grid(.~.sign)
```

## Table
```{r}
gse@result %>% as_tibble() %>% arrange(-abs(NES)) %>% filter(p.adjust < 0.05) %>%  DT::datatable()
```

# KEGG Pathway Enrichment
```{r, fig.width=3}
kk <- enrichKEGG(gene         = eg_diff_genes$ENTREZID, 
                 universe = eg_universe$ENTREZID,
                 organism     = 'hsa',
                 pvalueCutoff = 0.1)
dotplot(kk) + ggtitle("KEGG Pathway Enrichment") 
```

## Table
```{r}
kk <- setReadable(kk, OrgDb =  org.Hs.eg.db, keyType = 'ENTREZID')
kk@result %>% DT::datatable()
```



# Session Info
```{r}
devtools::session_info()
```