---
title: "Scavenger Hunt"
output: html_notebook
---

Find the Bharti (?) iRPE RNA-seq datasets (published and unpublished)

# ABCA4 KO (Mitra)
repo: abca4_ko_ipsc_rpe
meta: done
```{r}
library(tidyverse)
library(cowplot)
# data_dir <- '/Volumes/data/projects/nei/bharti/abca4_ko_ipsc_rpe__mitra/salmon_quant/'
# 
# nisc_meta <- read_tsv('../data/abca4_ko_ipsc_rpe/Brooks_ABCA4_RPE.report.txt') %>% as_tibble()
# metadata_abca4KO <- readxl::read_xlsx('../data/abca4_ko_ipsc_rpe/Brooks ABCA4_RPE  RNA seqIllumina_Sample_Submission_Form.xlsx', skip = 22, sheet = 'Sample Form') %>% 
#   filter(!is.na(`Sample ID ***`), !grepl('Tubes should', `Sample ID ***`) ) %>% 
#   dplyr::rename(Sample = 'Sample ID ***') %>% 
#   mutate(Treatment = case_when(grepl('\\+', Sample) ~ 'OS',
#                                TRUE ~ 'Control'),
#          ABCA4 = case_when(grepl('^C\\d\\d', Sample) ~ 'KO',
#                            TRUE ~ 'WT')) %>% 
#   mutate(Group = paste0(Treatment, '_', ABCA4)) %>% as_tibble()
# metadata_abca4KO <- metadata_abca4KO %>% 
#   mutate(submitted_name = gsub('-','_', `Sample ID  Validation`)) %>% 
#   left_join(nisc_meta) %>% 
#   rowwise() %>% 
#   mutate(file_path = gsub('_R[1|2]_001.fastq.gz','', file_name) %>% 
#            paste0(., '/quant.sf'),
#          SampleCore = str_split(Sample, '[+|\\s]')[[1]][1]) %>% 
#   mutate(SampleCore = case_when(SampleCore == 'STAPFKR' ~ 'STAP',
#                                 TRUE ~ SampleCore)) %>% 
#   dplyr::select(-file_name) %>% 
#   unique() %>% 
#   ungroup() %>% 
#   mutate(full_path = paste0(data_dir, file_path),
#          Owner = 'Mitra',
#          cell_line = SampleCore,
#          Sample = `Sample ID  Validation`,
#          Class = case_when(ABCA4 == 'WT' & Treatment == 'Control' ~ 'Control',
#                          TRUE ~ 'Perturbed')) 

# write_tsv(metadata_abca4KO, file = '../data/metadata_abca4KO.tsv')
metadata_abca4KO <- read_tsv('../data/metadata_abca4KO.tsv') %>% mutate(Repo = 'abca4_ko_ipsc_rpe')

metadata_abca4KO
```

# OCA (Brooks (Aman)) and Bharti (Ruchi, Karla))
repo: oca_rna-seq
meta: done
```{r}
#all sample / lanes
lane_metadata <- read_csv('../data/oca_rna-seq/sequence_metadata.csv')
lane_metadata <- lane_metadata  %>% rowwise() %>% mutate(lane_sample = str_split(file_name, '_R[1/2]')[[1]][1])
lane_metadata <- lane_metadata %>% 
  rowwise() %>% 
  mutate(Lane = str_split(lane_sample, '_')[[1]][4],
         FlowCell = str_split(lane_sample, '_')[[1]][1]) %>% 
  ungroup()


# Ruchi (HI/HIHS)
sample <- paste0(seq(1:21), "_RS")
treatment_hours <- rep(c(0,24,48,48,48,48,48),3)
cell_line <- c(rep('AMDCD',7), rep('3D1',7), rep('Z8', 7))
serum <- rep(c('None', 'HS', 'HS', 'HIHS', 'HS', 'HS', 'HS'), 3)
drug <- rep(c('None', 'None', 'None', 'None', 'Drug 1', 'Drug 3', 'Drug 4'), 3)
ruchi_metadata <- cbind(sample, cell_line, treatment_hours, serum, drug) %>% data.frame(stringsAsFactors = FALSE) %>% 
  mutate(DrugSimple = case_when(grepl('Drug', drug) ~ 'Drug',
                                TRUE ~ 'None'),
         Class = case_when(serum %in% c('None', 'HIHS') & drug == 'None' ~ 'Control',
                           TRUE ~ 'Perturbed'),
         Owner = 'Ruchi',
         CellType = 'iRPE') %>% 
  rename(Sample = sample,
         Treatment = drug)



# Aman (OCA)
rpe <- tibble::tribble(
  ~Name,     ~sample, ~Volume,   ~Concentration, ~Source,
  "OCA2",   "AG1", 10L, 158.2,   "OCA9 (SAMPLE1)",
  "OCA2",   "AG2", 10L, 108.5, "OCA103 (SAMPLE4)",
  "OCA2",   "AG3", 15L,  84.8,  "OCA71 (SAMPLE7)",
  "OCA1A",   "AG4", 10L,   143,  "OCA8 (SAMPLE10)",
  "OCA1A",   "AG5", 10L, 123.1, "OCA26 (SAMPLE13)")
fib <- tibble::tribble(
  ~sample, ~Genotype,
  "AMDCD1",                     "WT",
  "AMDCD2",                    "WT",
  "ATW1",                    "WT",
  "ATW2",                    "WT",
  "OTW1", "Compound heterozygous",
  "OTW2", "Compound heterozygous",
  "STW1",          "Heterozygous",
  "STW2",          "Heterozygous",
  "VCEQ1", "Compound heterozygous",
  "VCEQ2", "Compound heterozygous",
  "VED1",          "Heterozygous",
  "VED2",          "Heterozygous"
)
rpe$Cell <- 'iRPE'
fib$Cell <- 'Fibroblast'

aman_metadata <- bind_rows(rpe, fib) %>% 
  rename(CellType = Cell) %>% 
  mutate(cell_line = Source, 
         Owner = 'Aman',
         Class = case_when(Genotype == 'WT' ~ 'Control',
                           TRUE ~ 'Perturbed')) %>% 
  rename(Sample = sample)



# karla
karla_metadata <- rbind(
  c('KB#1','Ctrl shRNA Monolayer','Best 4C'),
  c('KB#2','Nox4 shRNA Monolayer','Best 4C'),
  c('KB#3','Intact Monolayer ','Best 4C'),
  c('KB#4','Ctrl shRNA Monolayer','Tuba 1B'),
  c('KB#5','Nox4 shRNA Monolayer','Tuba 1B'),
  c('KB#6','Intact Monolayer','Tuba 1B'),
  c('KB#7','Ctrl shRNA Trypsinized','Best 4C'),
  c('KB#8','Nox4 shRNA Trypsinized','Best 4C'),
  c('KB#9','Trypsinized','Best 4C'),
  c('KB#10','Ctrl shRNA Trypsinized','Tuba 1B'),
  c('KB#11','Nox4 shRNA Trypsinized','Tuba 1B'),
  c('KB#12','Trypsinized','Tuba 1B')) %>%
  data.frame() %>% 
  rename(sample = X1, Condition = X2, `Cell Line` = X3) %>% 
  mutate(sample=gsub('#', '', sample)) %>% 
  mutate(shRNA = case_when(grepl('Ctrl', Condition) ~ 'Ctrl', 
                           grepl('Nox4', Condition) ~ 'Nox4',
                           TRUE ~ 'None'),
         State = case_when(grepl('Monolayer', Condition) ~ 'Monolayer',
                           TRUE ~ 'Trypsinized'),
         Owner = 'Karla',
         CellType = 'iRPE',
         # even the "control" samples had a mock shRNA treatment
         Class = 'Perturbed') %>% 
  rename(cell_line = `Cell Line`,
         Sample = sample,
         Treatment = shRNA)





# bind all three
oca_meta <- bind_rows(karla_metadata, aman_metadata, ruchi_metadata) %>% mutate(Repo = 'oca_rna-seq')


oca_metadata <- lane_metadata %>% left_join(oca_meta, by = c('submitted_name' = 'Sample'))
```

# iRPE Dev Time Series (Irina)
repo: bharti_irina_iPSC_RPE_time_series-
meta: done
```{r}
irina_meta <- read_tsv('../data/bharti_irina_iPSC_RPE_time_series-/ids.txt') %>% 
  pivot_longer(-`Cell line`, names_to = 'Info', values_to = 'Sample') %>% 
  mutate(Note = case_when(grepl('dup', `Cell line`) ~ "Duplicate")) %>% 
  mutate(cell_line = gsub(' \\(.*', '', `Cell line`),
         CellType = case_when(grepl('iPS', Info) ~ 'iPSC',
                              TRUE ~ 'iRPE'),
         Class = NA,
         Owner = 'Irina',
         Repo = 'bharti_irina_iPSC_RPE_time_series-') 
```
# Peripheral - Fovea iRPE (Ortolan)
repo: ortolan_rpe_macula_peripheral_scRNA
meta: done
```{r}
ortolan_fq <- read_tsv('../data/ortolan_rpe_macula_peripheral/sample_table.tsv', col_names = c('info','fastq'))
ortolan_meta <- readxl::read_xlsx(path = '../data/ortolan_rpe_macula_peripheral/NISC Sample Submission Form_Bharti_DO.xlsx', sheet = 'Submission Form', skip = 21)
colnames(ortolan_meta)[9] <- 'info'
ortolan_meta <- ortolan_meta %>% rowwise() %>% 
  mutate(cell_line = str_split(info, ' ', simplify = TRUE)[1],
         info = gsub(' ','_',info)) %>% 
  filter(!is.na(`Sample ID***`), 
         !grepl('Tubes', `Sample ID***`)) %>% 
  mutate(Owner = 'Davide',
         CellType = 'iRPE',
         Class = case_when(grepl('DMSO', info) ~ 'Control',
                           TRUE ~ 'Perturbed'),
         Repo = 'ortolan_rpe_macula_peripheral_scRNA') %>% 
  rename(Sample = `Sample ID***`)

```
# Piezo1 KO iRPE (Qin)
repo: qin_piezo1_iRPE_2021
meta: done
```{r}
# working_dir <- '~/data/qin_piezo1_iRPE_2021/'
# files <- list.files(path=working_dir,recursive=TRUE,pattern='quant.sf', full.names = TRUE)
# 
# meta <- files %>% enframe() %>% 
#   mutate(sample = str_extract(value, 'Control_\\d|Drug_\\d'), 
#          treatment = str_extract(value, str_extract(value, 'Control|Drug')),
#          cell_line = 'D3C',
#          Class = case_when(grepl('Control', treatment) ~ 'Control',
#                            TRUE ~ 'Perturbed')) %>% 
#   rename(fastq_path = value)
# write_tsv(meta, file = '../data/qin_piezo1_iRPE_2021/meta.tsv')
qin_meta <- read_tsv('../data/qin_piezo1_iRPE_2021/meta.tsv') %>% 
  mutate(Owner = 'Qin',
         CellType = 'iRPE',
         Repo = 'qin_piezo1_iRPE_2021') %>% 
  rename(Sample = sample,
         Treatment = treatment)
```

# WNT CEP290 iRPE (Dominik)
repo: dominik_WNT_iRPE_2021
meta: done
```{r}
dominik_meta <- readxl::read_xlsx('../data/dominik_WNT_iRPE_2021/Layout experiment.xlsx') %>% 
  rename(Sample = `Sample name`,
         cell_line = Cells) %>% 
  mutate(Class = case_when(Treatment == 'CTRL'  & !grepl("CEP", cell_line)~ 'Control',
                           TRUE ~ 'Perturbed'),
         Owner = 'Dominik',
         CellType = 'iRPE',
         Repo = 'dominik_WNT_iRPE_2021')

```

# Merge all
```{r}
ogvfb_meta <- bind_rows(dominik_meta, 
          qin_meta,
          ortolan_meta,
          irina_meta,
          oca_meta,
          metadata_abca4KO) %>% 
  relocate(Sample, cell_line, CellType, Treatment, Owner, Class) %>% 
  mutate(Class = case_when(Sample %in% c('KB5','KB9','KB10','KB12','2_RS','3_RS','9_RS','10_RS','16_RS','17_RS') ~ 'Degeneration',
         TRUE ~ Class))

write_tsv(ogvfb_meta, file = '../data/ogvfb_full_meta.tsv')

write_csv(ogvfb_meta %>% filter(Class == 'Control', CellType == 'iRPE'), file = '../data/control_meta_iRPE.csv')
write_csv(ogvfb_meta %>% filter(Class == 'Control', CellType != 'iRPE'), file = '../data/control_meta_notRPE.csv')

write_csv(ogvfb_meta %>% filter(Class != 'Control', CellType == 'iRPE'), file = '../data/perturbed_meta_iRPE.csv')
write_csv(ogvfb_meta %>% filter(Class != 'Control', CellType != 'iRPE'), file = '../data/perturbed_meta_notRPE.csv')
```

# Quick Plots
```{r, fig.width=6, fig.height=8}
a <- ogvfb_meta %>% ggplot(aes(x=Class)) + geom_bar(position = position_dodge()) + 
  cowplot::theme_cowplot() +
  facet_wrap(~CellType) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
b <- ogvfb_meta %>% ggplot(aes(x=Class, fill = Owner)) + geom_bar(position = position_dodge()) + 
  cowplot::theme_cowplot() +
  scale_fill_manual(values = pals::glasbey() %>% unname()) +
  facet_wrap(~CellType) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))
c <- ogvfb_meta %>% 
  mutate(Line = case_when(grepl('OCA', cell_line) ~ 'OCA Patient',
                          cell_line == 'Best 4C' ~ 'BEST4C',
                          TRUE  ~ cell_line)) %>% 
  ggplot(aes(x=Line, fill = Class)) + geom_bar(position = position_dodge()) + 
  cowplot::theme_cowplot() +
  scale_fill_manual(values = pals::alphabet2() %>% unname()) +
  coord_flip()+
  facet_wrap(~CellType)

ab <- cowplot::plot_grid(a, b, ncol = 2)
cowplot::plot_grid(ab, c, ncol = 1, rel_heights = c(0.6,1))

```

